lib/streamlit/uploaded_file_manager.py

```diff
@@ -72,67 +78,127 @@ def __init__(self):
+        widget_id
+            The widget ID of the FileUploader that created the file.
         """
-        self._add_files(session_id, widget_id, files)
-        self.on_files_updated.send(session_id)
+        file_list_id = (session_id, widget_id)
+        with self._files_lock:
+            return self._files_by_id.get(file_list_id, []).copy()
 
     def get_files(
-        self, session_id: str, widget_id: str
-    ) -> Optional[List[UploadedFileRec]]:
-        """Return the file list with the given ID, or None if the ID doesn't
-        exist.
+        self, session_id: str, widget_id: str, file_ids: List[int]
+    ) -> List[UploadedFileRec]:
+        """Return the files with the given widget_id and file_ids.
 
         Parameters
         ----------
-        session_id : str
+        session_id
             The session ID of the report that owns the file.
-        widget_id : str
+        widget_id
             The widget ID of the FileUploader that created the file.
+        file_ids
+            List of file IDs. Only files whose IDs are in this list will be
+            returned.
+        """
+        return [
+            f for f in self.get_all_files(session_id, widget_id) if f.id in file_ids
+        ]
 
-        Returns
-        -------
-        list of UploadedFileRec or None
+    def remove_orphaned_files(
+        self,
+        session_id: str,
+        widget_id: str,
+        newest_file_id: int,
+        active_file_ids: List[int],
+    ) -> None:
+        """Remove 'orphaned' files: files that have been uploaded and
+        subsequently deleted, but haven't yet been removed from memory.
+
+        Because FileUploader can live inside forms, file deletion is made a
+        bit tricky: a file deletion should only happen after the form is
+        submitted.
+
+        FileUploader's widget value is an array of numbers that has two parts:
+        - The first number is always 'this.state.newestServerFileId'.
+        - The remaining 0 or more numbers are the file IDs of all the
+          uploader's uploaded files.
+
+        When the server receives the widget value, it deletes "orphaned"
+        uploaded files. An orphaned file is any file associated with a given
+        FileUploader whose file ID is not in the active_file_ids, and whose
+        ID is <= `newestServerFileId`.
+
+        This logic ensures that a FileUploader within a form doesn't have any
+        of its "unsubmitted" uploads prematurely deleted when the script is
+        re-run.
         """
         file_list_id = (session_id, widget_id)
         with self._files_lock:
-            return self._files_by_id.get(file_list_id, None)
+            file_list = self._files_by_id.get(file_list_id)
+            if file_list is None:
+                return
+
+            # Remove orphaned files from the list
+            new_list = [
+                f for f in file_list if f.id > newest_file_id or f.id in active_file_ids
```

Comment:
Making sure I understand this.

f.id > newest_file_id = "a new file was uploaded here"
f.id in active_file_ids = "a file was sent by the widget itself"?

Reply:
Yeah:

f.id in active_file_ids = "this file is currently being tracked by the widget"
f.id > newest_file_id = "this file was uploaded *after* the the widget was most recently updated. probably because it's in a form."
f.id < newest_file_id && f.id not in active_file_ids = "this file is not currently tracked by the widget, and was not uploaded after the most recent update. So it's been deleted by the user on the frontend, and is now "orphaned", so the server needs to clean it up to get parity"

Reply:
(I'll add this to the code comment, to make it clear to future readers.)